### PR TITLE
[Sampler] Introduce logprobs mode for logging

### PR DESCRIPTION
**Pull request by houseroad**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose
So we can specify logprobs mode to log different content in the logprobs in the response.
We have four modes supported at this moment, "raw_logprobs", "processed_logprobs", "raw_logits", "processed_logits"

We use model config instead of SampleParams since we don't think there is real use case to mix things in the serving instance. We would like to simplify the logits for now.

## Test Plan
```
CUDA_VISIBLE_DEVICES=1 HF_HUB_DISABLE_XET=1 with-proxy pytest tests/v1/sample/test_logprobs.py::test_logprobs_mode
```

## Test Result
All 4 tests passed.

## (Optional) Documentation Update
Added

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -316,6 +316,13 @@ class ModelConfig:
     """Maximum number of log probabilities to return when `logprobs` is
     specified in `SamplingParams`. The default value comes the default for the
     OpenAI Chat Completions API."""
+    logprobs_mode: str = "processed_logprobs"
```
Comment (22quinn) [inline vllm/config.py]: maybe use `Literal` type to restrict the choices?

File: vllm/config.py [hunk 2]
```diff
@@ -316,6 +316,13 @@ class ModelConfig:
...
+    """Indicates the content returned in the logprobs and prompt_logprobs.
+    Supported mode:
+    1) raw_logprobs, 2 processed_logprobs, 3) raw_logits, 4) processed_logits.
```
Comment (simon-mo) [inline vllm/config.py]:
```suggestion
    1) raw_logprobs, 2) processed_logprobs, 3) raw_logits, 4) processed_logits.
```

File: tests/v1/sample/test_logprobs.py [hunk 3]
```diff
@@ -426,3 +426,44 @@ def test_zero_logprobs(vllm_model, example_prompts,
...
             assert len(prompt_token_ids) == len(prompt_logprobs)
+
+
+@pytest.mark.parametrize(
+    "logprobs_mode",
+    ["raw_logprobs", "raw_logits", "processed_logprobs", "processed_logits"])
+def test_logprobs_mode(logprobs_mode: str, monkeypatch: pytest.MonkeyPatch):
+    """vLLM v1 engine should fail a request with `logprobs > max_logprobs`
+    Should also fail for `prompt_logprobs > max_logprobs`
+    APC should not matter as this test checks basic request validation.
```
Comment (simon-mo) [inline tests/v1/sample/test_logprobs.py]: this comment is stale?

**Review by simon-mo — APPROVED** [review]

File: tests/v1/sample/test_logprobs.py (hunk 3 above)
Comment (houseroad) [reply tests/v1/sample/test_logprobs.py]: Good catch.

File: vllm/v1/sample/sampler.py [hunk 4]
```diff
@@ -36,7 +38,10 @@ def forward(
         # See https://vllm-dev.slack.com/archives/C07UUL8E61Z/p1735907856007919 # noqa: E501
         num_logprobs = sampling_metadata.max_num_logprobs
         if num_logprobs is not None:
-            raw_logprobs = self.compute_logprobs(logits)
+            if self.logprobs_mode == "raw_logprobs":
+                raw_logprobs = self.compute_logprobs(logits)
+            elif self.logprobs_mode == "raw_logits":
+                raw_logprobs = logits.clone()
```
Comment (yeqcharlotte) [inline vllm/v1/sample/sampler.py:44]: nit: raw logits, processed logits/logprobs are all called `raw_logprobs` :)

File: tests/v1/sample/test_logprobs.py [hunk 5]
```diff
@@ -426,3 +427,45 @@ def test_zero_logprobs(vllm_model, example_prompts,
...
+                       monkeypatch: pytest.MonkeyPatch):
+    """Test with LLM engine with different logprobs_mode.
+    For logprobs, we should have non-positive values.
+    For logits, we should expect at least one positive values.
+    """
+    from vllm import LLM
+    with monkeypatch.context() as m:
+        m.setenv("VLLM_USE_V1", "1")
+
+        llm = LLM(
```
Comment (yeqcharlotte) [inline tests/v1/sample/test_logprobs.py:445]: does this test run on cpu by default?

Comment (houseroad) [reply tests/v1/sample/test_logprobs.py:445]: it should be on GPU by default.

**Comment by youkaichao** [comment]
@houseroad @22quinn maybe we should update the doc https://docs.vllm.ai/en/latest/usage/v1_guide.html?h=immediately#semantic-changes-to-logprobs about the logprobs mode argument?

**Comment by 22quinn** [comment]
> @houseroad @22quinn maybe we should update the doc https://docs.vllm.ai/en/latest/usage/v1_guide.html?h=immediately#semantic-changes-to-logprobs about the logprobs mode argument?

SG, will update this and probably add a detailed doc on Sampler. I spent some time tracing through the whole sampling process and feel it could be unclear/confusing for general users.
